### PR TITLE
Add MakeSDPA attention

### PR DESCRIPTION
**Pull request by CuriousPanCake**
Add MakeSDPA attention

Add MakeSDPA attention transformation that inserts a ScaledDotProductAttention node for the future processing in SDPAToPagedAttention.

Previously this was done via Optimum-Intel

Currently supported models:
	* katuni4ka/tiny-random-jais

### Details:
 - *item1*
 - *...*

### Tickets:
 - CVS-138933

### CODE REVIEW

File: src/core/src/pass/make_sdpa.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    auto add_label = pattern::wrap_type<op::v1::Add>({select_label, pattern::any_input()});
+    auto add_label1 = pattern::wrap_type<op::v1::Add>({add_label, pattern::any_input()});
+
+    auto softmax_label = pattern::wrap_type<op::v8::Softmax>({add_label1});
+    auto V_label = pattern::any_input();
+    auto final_matmul_label = pattern::wrap_type<op::v0::MatMul>({softmax_label, V_label});
+
+    matcher_pass_callback callback = [OV_CAPTURE_CPY_AND_THIS](pattern::Matcher& m) {
+        std::cout << "____" << matcher_name << "____" << std::endl;
```
Comment (rkazants) [inline src/core/src/pass/make_sdpa.cpp:40]:
```suggestion
```

File: src/core/src/pass/make_sdpa.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        auto scale = pattern_map.at(scale_label);
+        auto scale_converted = std::make_shared<op::v1::Divide>(op::v0::Constant::create(element::f32, Shape{}, {1}), scale);
+        auto dummy_mask = op::v0::Constant::create(element::i32, Shape{}, {0});
+
+        auto sdpa = std::make_shared<ov::op::v13::ScaledDotProductAttention>(Q, K, V, dummy_mask, scale_converted, true);
+        sdpa->set_friendly_name(m.get_match_root()->get_friendly_name());
+        ov::replace_node(m.get_match_root(), sdpa);
+
+        if (transformation_callback(sdpa)) {
+            std::cout << "transformation_callback() == false" << std::endl;
```
Comment (rkazants) [inline src/core/src/pass/make_sdpa.cpp:56]:
```suggestion
```

**Review by rkazants — COMMENTED** [review]
may we have tests and comments with explanation (why it is equivalent to paged attention)?

**Comment by CuriousPanCake** [comment]
> may we have tests and comments with explanation (why it is equivalent to paged attention)?

This PR is a part of supporting PagedAttention transformation for LLMs. Right now, I expect more changes here to come.

File: src/core/include/openvino/pass/make_sdpa.hpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace ov {
+namespace pass {
+
+class TRANSFORMATIONS_API MakeSDPA;
+
+}  // namespace pass
+}  // namespace ov
+
+class ov::pass::MakeSDPA : public ov::pass::MatcherPass {
```
Comment (itikhono) [inline src/core/include/openvino/pass/make_sdpa.hpp:18]: it's better to rename it to SDPAFusion

File: src/core/src/pass/make_sdpa.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            return false;
+        }
+
+        ov::copy_output_runtime_info(
```
Comment (itikhono) [inline src/core/src/pass/make_sdpa.cpp:60]: please use `ov::copy_runtime_info`

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp [hunk 5]
```diff
@@ -23,21 +25,59 @@ ov::pass::TotalSequenceLengthPattern::TotalSequenceLengthPattern(
...
-        auto required_shape = gather->get_output_partial_shape(0);
-        if (replacement->get_output_partial_shape(0) != required_shape && required_shape.rank().is_static()) {
-            replacement = op::util::reshapeTo(replacement, Shape(required_shape.rank().get_length(), 1));
+        auto gather_axis_data = gather_axis->get_data_ptr<int32_t>();
+
+        const auto& gather_data_shape = gather->input(0).get_shape();
+
+        std::shared_ptr<Node> replacement = max_context_len;
+
+        if (concat->get_axis() == static_cast<int64_t>(gather_data_shape[*gather_axis_data] - std::abs(*gather_idx_data))) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp:62]: we should calculate gather_data_shape[*gather_axis_data] - std::abs(*gather_idx_data) if gather_idx_data is negative, for positive we can use the value itself
moreover ShapeOf always returns 1D tensor, so gather_axis_data should be always zero, so it's safe to use 0

File: src/core/src/pass/make_sdpa.cpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+        auto Q = pattern_map.at(Q_label);
+        auto K = pattern_map.at(K_label);
+        auto V = pattern_map.at(V_label);
+
+        auto scale = pattern_map.at(scale_label);
+        auto scale_converted = std::make_shared<op::v1::Divide>(op::v0::Constant::create(element::f32, Shape{}, {1}), scale);
+        auto dummy_mask = op::v0::Constant::create(element::i32, Shape{}, {0});
+
+        auto sdpa = std::make_shared<ov::op::v13::ScaledDotProductAttention>(Q, K, V, dummy_mask, scale_converted, true);
```
Comment (praasz) [inline src/core/src/pass/make_sdpa.cpp:50]: The `ov::pass::NodeRegistry` can be used to create some nodes, Then copy runtime info can be simpler e.g.
`copy_runtime_info(from, registry.get());`

File: src/core/src/pass/sdpa_to_paged_attention.cpp [hunk 7]
```diff
@@ -31,9 +32,9 @@ static std::shared_ptr<v0::Parameter> setName(std::shared_ptr<v0::Parameter> nod
 bool ov::pass::SDPAToPagedAttention::run_on_model(const std::shared_ptr<ov::Model>& model) {
     RUN_ON_MODEL_SCOPE(SDPAToPagedAttention);
 
-    OPENVINO_ASSERT(ov::op::util::has_op_with_type<ov::op::v13::ScaledDotProductAttention>(model),
-                    "No ScaledDotProductAttention operation observed in the graph, cannot perform"
-                    "the SDPAToPagedAttention transformation.");
+    // OPENVINO_ASSERT(ov::op::util::has_op_with_type<ov::op::v13::ScaledDotProductAttention>(model),
+    //                 "No ScaledDotProductAttention operation observed in the graph, cannot perform"
+    //                 "the SDPAToPagedAttention transformation.");
```
Comment (praasz) [inline src/core/src/pass/sdpa_to_paged_attention.cpp:37]:
```suggestion
```
or uncomment it?

File: src/core/include/openvino/pass/make_sdpa.hpp [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+#pragma once
+
+#include "openvino/pass/graph_rewrite.hpp"
+#include "transformations_visibility.hpp"
+
+namespace ov {
+namespace pass {
+
+class TRANSFORMATIONS_API MakeSDPA;
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/pass/make_sdpa.hpp:13]: is it required to be in public headers?

Comment (CuriousPanCake) [reply src/core/include/openvino/pass/make_sdpa.hpp:13]: @ilya-lavrenov yes, we added them to public headers for linking with tests which we added recently

Comment (ilya-lavrenov) [reply src/core/include/openvino/pass/make_sdpa.hpp:13]: public API is required for users, not for tests

**Pull request closed without merging**